pkg/v1beta1/challengesolver/solver_test.go

```diff
@@ -0,0 +1,210 @@
+package challengesolver_test
+
+import (
+	"context"
+	"fmt"
+	"hash/adler32"
+	"testing"
+
+	acmev1 "github.com/cert-manager/cert-manager/pkg/apis/acme/v1"
+	"github.com/kanopy-platform/gateway-certificate-controller/pkg/v1beta1/cache"
+	"github.com/kanopy-platform/gateway-certificate-controller/pkg/v1beta1/challengesolver"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/stretchr/testify/assert"
+
+	certmanagerfake "github.com/cert-manager/cert-manager/pkg/client/clientset/versioned/fake"
+	acmefake "github.com/cert-manager/cert-manager/pkg/client/clientset/versioned/typed/acme/v1/fake"
+	networkingv1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+	istiofake "istio.io/client-go/pkg/clientset/versioned/fake"
+	networkingv1beta1fake "istio.io/client-go/pkg/clientset/versioned/typed/networking/v1beta1/fake"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	corev1fake "k8s.io/client-go/kubernetes/typed/core/v1/fake"
+	k8stesting "k8s.io/client-go/testing"
+)
+
+func TestStub(t *testing.T) {
+	t.Parallel()
+
+	glc := cache.New()
+
+	assert.NotNil(t, glc)
+}
+
+type testHelper struct {
+	ics *istiofake.Clientset
+	ccs *certmanagerfake.Clientset
+	scs *k8sfake.Clientset
+	glc *cache.GatewayLookupCache
+}
+
+func (th *testHelper) newTestSolver() *challengesolver.ChallengeSolver {
+	return challengesolver.NewChallengeSolver(th.scs.CoreV1(), th.ics.NetworkingV1beta1(), th.ccs.AcmeV1(), th.glc)
+}
+
+func TestChallengeSolver(t *testing.T) {
+	for _, test := range []struct {
+		name           string
+		challenge      *acmev1.Challenge
+		service        corev1.Service
+		virtualService *networkingv1beta1.VirtualService
+		gatewayName    string
+		pass           bool
+		validateVS     bool
+		noRequeue      bool
+	}{
+		{
+			name:      "no challenge",
+			pass:      true,
+			noRequeue: true,
+		},
+		{
+			name:      "No Gateway",
+			challenge: getChallenge("noservice", "example", "noservice.com"),
+			pass:      false,
+		},
+		{
+			name:      "No Gateway",
+			challenge: getChallenge("noservice", "example", "noservice.com"),
+			pass:      false,
+		},
+		{
+			name:        "No Service",
+			challenge:   getChallenge("noservice", "example", "noservice.com"),
+			gatewayName: "gateway",
+			pass:        false,
+		},
+		{
+			name:        "No Service Port",
+			challenge:   getChallenge("noservice", "example", "noservice.com"),
+			gatewayName: "gateway",
+			service:     getService("noportservice", "example", "noportservice.com", 0),
+			pass:        false,
+		},
+		{
+			name:        "Service",
+			challenge:   getChallenge("noservice", "example", "service.com"),
```

Comment:
nit: This argument in getChallenge is probably unnecessary.
```suggestion
			challenge:   getChallenge("service", "example", "service.com"),
```